Validator: Scientific notation and NA values

sandertan opened the pull request:
# What? Why?
Add some possible scenarios that should not error. When recreating staging files for the CCLE study, I noticed some values are causing errors in the validator. The importer does not complain about them. As far as I noticed, these values do not break anything.

The first fix is for when a chromosomal location is written in scientific notation, such as `115,000,000` written as `1.15e+08` in the CNA segmented data:

![screen shot 2017-03-16 at 10 32 44](https://cloud.githubusercontent.com/assets/9624990/23989856/5f8bf2d2-0a34-11e7-9764-ecb3b8f5efe6.png)

The second fix affects a couple of optional columns in the mutations data, after running Maf2Maf:

![screen shot 2017-03-16 at 10 32 55](https://cloud.githubusercontent.com/assets/9624990/23989857/5fa82b32-0a34-11e7-9014-43df4ba0ea26.png)

fedde-s commented:
@sandertan, did you test if the importer does indeed

- [ ] insert the same value into the database when the `[nt]_(alt|ref)_count` columns of mutation data files contain a `.` as when they are blank,
- [ ] insert the same value when the columns`Validation_Status` and `Verification_Status` contain the string `NA` as when they are blank
- [ ] insert the expected integer value when  a scientific-notation floating point string with a `+` sign occurs as a genomic start or end coordinate in segment-wise CNA data files, and
- [ ] do the same for genomic coordinates in GISTIC_GENES data files?

Especially for the mutation data, I don't think we can make too many assumptions on whether servlets, API logic and front-end code will post-harmonise these values for us if we load invalid values into the database.

fedde-s commented:
If these values are not treated identically but they definitely should in all cases be equivalent, we should add the same logic to the importer code before merging this.

sandertan commented:
Thanks Fedde, I'll test it.

fedde-s commented on core/src/main/scripts/importer/validateData.py:
```diff
@@ -1895,6 +1895,9 @@ def checkLine(self, data):
+                    value = float(value.replace("e+", "e"))
                 try:
                     parsed_coords[col_name] = int(value)
                 except ValueError:
```
Another issue here: please log the appropriate message about the validity of the data file format if this field is not a floating-point number instead of letting the validation script crash with an exception, and also log a message if the floating-point number that does not make sense as a genomic coordinate.

fedde-s commented on core/src/main/scripts/importer/validateData.py:
```diff
@@ -2435,6 +2438,9 @@ def parse_genomic_coord(self, value, column_number):
+            value = float(value.replace("e+", "e"))
         try:
             parsed_value = int(value)
         except ValueError:
```
The same issue as in segment-wise CNA data files here: please log the appropriate message about the validity of the data file format if this field is not a floating-point number instead of letting the validation script crash with an exception, and also log a message if the floating-point number that does not make sense as a genomic coordinate.

sandertan commented:
As @fedde-s suggested, I tested with importing with different values in the `[nt]_(alt|ref)_count` columns. The importer writes `-1` when they are blank or `.`, so this validator change is okay.

When `Validation_Status` and `Verification_Status` are blank or `NA`, the importer writes `NA` to database. 

![screen shot 2017-03-16 at 14 09 46](https://cloud.githubusercontent.com/assets/9624990/23997781/1f031640-0a53-11e7-814c-ed442714377b.png)

I also tested the segment-wise CNA data files, the database seems to import `1.15e+08` correctly as `115,000,000`. 

![screen shot 2017-03-16 at 12 06 26](https://cloud.githubusercontent.com/assets/9624990/23997791/27925f6e-0a53-11e7-8c9e-614073914e7c.png)

I don't have test data for GISTIC peaks, so I removed this change from the commit.

fedde-s commented on core/src/main/scripts/importer/validateData.py:
```diff
@@ -1896,6 +1896,11 @@ def checkLine(self, data):
+                    if "e+" in value:
+                        value = float(value.replace("e+", "e"))
+                        if int(value) == value:
+                            pass
```
This conditional does not seem to do anything in either case. Could be useful raise a ValueError or log something if [value.is_integer()](https://docs.python.org/2/library/stdtypes.html#float.is_integer) returns false?

sandertan replied:
Done

sandertan commented:
Before merging, I see if I can add '---' as empty value for `Verification_Status`, which is causing an error in the skcm_tcga study, reported in https://github.com/cBioPortal/datahub/issues/4

sandertan commented:
Addition to this PR:

In `skcm_tcga` the `Validation_Status` contains the value `---` 2677 times. The importer imports this as `---`:

![screen shot 2017-04-04 at 13 20 02](https://cloud.githubusercontent.com/assets/9624990/24656608/7de51420-1943-11e7-88a3-d579a5a1bbbe.png)


This screenshot shows the frontend does not have troubles with this value:

![screen shot 2017-04-04 at 13 21 18](https://cloud.githubusercontent.com/assets/9624990/24656609/7e0a4d1c-1943-11e7-98d1-05514128315e.png)


[Link](http://www.cbioportal.org/index.do?cancer_study_list=skcm_tcga&cancer_study_id=skcm_tcga&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=skcm_tcga_mutations&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=skcm_tcga_sequenced&case_ids=&patient_case_select=sample&gene_set_choice=user-defined-list&gene_list=CACNA1D&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit&show_samples=false&) to the public portal for this study - gene - samples (sort by AA change to get samples with `Validations Status ---` on top).

So I added `---` as valid value for `Validation_Status`. However, this value is not desribed by cBioPortal [FileFormat](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md) (`--` sort of is). In addition, this is not a valid value according to the [NCI MAF specification](https://wiki.nci.nih.gov/display/TCGA/Mutation+Annotation+Format+%28MAF%29+Specification). Last, I'm not sure how this value got in this datafile and therefor it might be better to rerun maf2maf to regenerate this file..

fedde-s commented:
I did  not review the latest addition to this pull request. @pieterlukasse, do you have time for that?

pieterlukasse commented:
The changes proposed here look fine to me. Only one concern regarding the `---` value being added to the DB: I would like to avoid that. Can this be converted to `NA` by the loader? Is `---` the same as `NA`? I would propose we store both as `NA` in the DB in this case.
@zheins @n1zea144 Can you check if it would be correct to interpret `Validation_Status=---` as being `Validation_Status=NA`?

n1zea144 commented:
@pieterlukasse The importing code in the cBioPortal looks at validation status to filter out events ([MutationFilter](https://github.com/cBioPortal/cbioportal/blob/master/core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java#L128)). It will filter out events if the validation status is "Redacted".  This was a convention worked among some groups at MSK, it is not a convention that came out of TCGA.  According to the TCGA MAF Spec the following values are permitted:

- Untested
- Inconclusive
- Valid
- Invalid

If you see a '---' as a validation status, it was probably a bug in the firehose output.  I would update the MAF and replace the values with NA.

sandertan commented:
@n1zea144 Thanks for the comment. 

I've removed the commit that would allow `---` and removed these values from the Validation_Status column in the MAF file of `skcm_tcga` in https://github.com/cBioPortal/datahub/pull/39 .
